Fix multi master sync bug

righdforsa opened the pull request:
Hey guys, round two of my fix for the bug where bedrock has the potential to end up with no leader because the primary leader is still syncing but because the secondary leader sees it with a higher priority the secondary leader won't stand up.

The solution is to use a priority of "-1" until a node reaches waiting the first time (i.e. finishes syncing)

In order to prevent this from breaking permafollowers on login, I've removed the checks that refuse login for permafollowers who aren't priority zero and non-permafollowers who are. At worst this opens us up to misconfigurations which shouldn't happen.

Along the way I fixed a "too fast leader flip flop" issue by clearing out the upgrade DB state for nodes that become leader and then lose quorum before it can be replicated. Having more nodes at priority -1 probably made it possible.

I've also tagged the test name in the "cout" output in any the cluster tests that had it.

Let's rock.

righdforsa commented:
Found another bug related to 0 priority:
```
2020-03-16T21:22:50.959370+00:00 expensidev bedrock10016: xxxxxx (SQLiteNode.cpp:551) update [sync] [dbug] {cluster_node_5/WAITING} Dumping evaluated cluster state: numLoggedInFullPeers=3 freshestPeer=cluster_node_0 highestPriorityPeer=cluster_node_0 currentLeader=none
2020-03-16T21:22:50.959432+00:00 expensidev bedrock10016: xxxxxx (SQLiteNode.cpp:584) update [sync] [eror] {cluster_node_5/WAITING} Assertion failed: (_priority > 0) != true
2020-03-16T21:22:50.959993+00:00 expensidev bedrock10016: xxxxxx (SLog.cpp:15) SLogStackTrace [sync] [warn]
2020-03-16T21:22:50.960078+00:00 expensidev bedrock10016: xxxxxx (SLog.cpp:15) SLogStackTrace [sync] [warn] SLogStackTrace() [0x693ce8]
...
```

righdforsa commented:
Latest changes passed 100 out of 100 times (failed once in the last pass, so technically, it passed 199 times out of 200)
```
[ TEST RESULTS ] Passed: 99, Failed: 0
[==============]
[       PASSED ] MultipleLeaderSyncTest::test
[==============]
[ TEST RESULTS ] Passed: 100, Failed: 0
[==============]
```

Rerunning all the other tests in Bedrock and Auth to make sure that nothing new broke. 🤞

righdforsa commented:
Bedrock tests:
```
[==============]
[ TEST RESULTS ] Passed: 129, Failed: 0
[==============]
```
Bedrock cluster tests:
```
[==============]
[ TEST RESULTS ] Passed: 25, Failed: 0
[==============]
```
Auth tests:
```
[==============]
[ TEST RESULTS ] Passed: 1191, Failed: 0
[==============]
```

All good, ready to roll.

righdforsa commented:
Bump. Ready for review, guys.

righdforsa commented:
> I don't actually think your worst case is right. The worst case is a node gets counted for or not count for quorum by master and can approve a transaction it shouldn't be able to approve.

Yes, you are correct, a member of the worst case class of "misconfigurations that shouldn't happen" is "half configured to be a permafollower", which could mean:
a) expected to be a permafollower but counted towards quorum (permanently removing one node from participating in quorum), or
b) expected to be a full peer, approving transactions, but not counted towards quorum, possible, in rare cluster circumstances, to become master

I'm more worried about a, but either way, when I mean "shouldn't" happen, I mean shouldn't happen because the production configs are pretty stable. Of course, that's no comfort if we can have a misconfiguration hiding out for a long time before we discover it. 

So, alternatively, for the login test we could use "Permafollower" instead of priority, or we could login with both "originalPriority" and "currentPriority". I'd be happy to make either of those changes, preference to the first one.

I seem to recall another option, but it's escaping me for the moment.

coleaeason commented:
>I'm more worried about a, but either way, when I mean "shouldn't" happen, I mean shouldn't happen because the production configs are pretty stable. Of course, that's no comfort if we can have a misconfiguration hiding out for a long time before we discover it.

Well, we actually hit these conditions every time we change the status of a permafollower. We hit these log lines thousands of times, indicating that if you remove them we would be falling through to the conditions we are specifically worried about. I don't think that's acceptable, so I think we should look at one of your other suggested fixes.

righdforsa commented:
Ok, I have replaced those conditions and added a Permafollower test to catch permafollower issues in the future. I haven't created a "intentionally misconfigured nodes" test, but in the course of my testing, I did break the LOGIN message in ways that give me confidence in the updated version of the conditionals. e.g. when I had the logic of the "Permafollower" parameter in the login message inverted, (full peers were sending "permafollower = true") the clustertests would fail since they couldn't establish quorum. Let me know if you want something more substantial than that @coleaeason. Currently running the new permafollower test 100 times to confirm test stability, should be done in 30-45 mins.

righdforsa commented:
After about 43 tests, I had a crash. I believe this is unrelated to my changes, but I've seen it before, too. Might be another thing worth fixing:
```
2020-04-01T22:10:10.613288+00:00 expensidev bedrock10001: xxxxxx (SQLite.cpp:472) verifyTable [sync] [info] 'test' already exists with correct schema.
2020-04-01T22:10:10.613440+00:00 expensidev bedrock10001: xxxxxx (libstuff.cpp:2449) SQuery [sync] [dbug] ROLLBACK
2020-04-01T22:10:10.613889+00:00 expensidev bedrock10001: xxxxxx (BedrockServer.cpp:2026) _upgradeDB [sync] [info] Finished running DB upgrade.
2020-04-01T22:10:10.614243+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:164) _SSignal_StackTrace [sync] [warn] Signal Segmentation fault(11) caused crash, logging stack trace.
2020-04-01T22:10:10.614499+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn]
2020-04-01T22:10:10.614673+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] _SSignal_StackTrace(int, siginfo_t*, void*) [0x6704cc]
2020-04-01T22:10:10.614844+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390) [0x7fc221bb4390]
2020-04-01T22:10:10.615005+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] std::basic_ostream<char, std::char_traits<char> >& std::operator<< <char, std::char_traits<char>, std::allocator<char> >(std::basic_ostream<char, std::char_traits<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [0x7fc2216e7200]
2020-04-01T22:10:10.615163+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] BedrockServer::sync(SData const&, std::atomic<STCPNode::State>&, std::atomic<bool>&, std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&, BedrockTimeoutCommandQueue&, BedrockServer&) [0x5335ba]
2020-04-01T22:10:10.615327+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] BedrockServer::syncWrapper(SData const&, std::atomic<STCPNode::State>&, std::atomic<bool>&, std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&, BedrockTimeoutCommandQueue&, BedrockServer&) [0x52f542]
2020-04-01T22:10:10.615483+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] void std::_Bind_simple<void (*(std::reference_wrapper<SData const>, std::reference_wrapper<std::atomic<STCPNode::State> >, std::reference_wrapper<std::atomic<bool> >, std::reference_wrapper<std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::reference_wrapper<BedrockTimeoutCommandQueue>, std::reference_wrapper<BedrockServer>))(SData const&, std::atomic<STCPNode::State>&, std::atomic<bool>&, std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&, BedrockTimeoutCommandQueue&, BedrockServer&)>::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul, 5ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul, 5ul>) [0x58648b]
2020-04-01T22:10:10.615785+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] std::_Bind_simple<void (*(std::reference_wrapper<SData const>, std::reference_wrapper<std::atomic<STCPNode::State> >, std::reference_wrapper<std::atomic<bool> >, std::reference_wrapper<std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::reference_wrapper<BedrockTimeoutCommandQueue>, std::reference_wrapper<BedrockServer>))(SData const&, std::atomic<STCPNode::State>&, std::atomic<bool>&, std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&, BedrockTimeoutCommandQueue&, BedrockServer&)>::operator()() [0x585d36]
2020-04-01T22:10:10.616094+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] std::thread::_State_impl<std::_Bind_simple<void (*(std::reference_wrapper<SData const>, std::reference_wrapper<std::atomic<STCPNode::State> >, std::reference_wrapper<std::atomic<bool> >, std::reference_wrapper<std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::reference_wrapper<BedrockTimeoutCommandQueue>, std::reference_wrapper<BedrockServer>))(SData const&, std::atomic<STCPNode::State>&, std::atomic<bool>&, std::atomic<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >&, BedrockTimeoutCommandQueue&, BedrockServer&)> >::_M_run() [0x58548a]
2020-04-01T22:10:10.616457+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xd0840) [0x7fc221676840]
2020-04-01T22:10:10.616668+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7fc221baa6ba]
2020-04-01T22:10:10.616831+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:167) _SSignal_StackTrace [sync] [warn] /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7fc2210cb41d]
2020-04-01T22:10:10.616989+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:171) _SSignal_StackTrace [sync] [warn] Calling DIE function.
2020-04-01T22:10:10.617149+00:00 expensidev bedrock10001: xxxxxx (SSignal.cpp:174) _SSignal_StackTrace [sync] [warn] DIE function returned, aborting (if not done).
```

righdforsa commented:
Ok, updated, again. Found some crash conditions in BedrockServer.cpp and protected against them. Ran my permafollower test with 1000 repeats and it passed. Ran clustertest 4 times with 10 repeats and it passed with zero failures twice. (I don't know if this is better/worse/same than current success rates)

Now Bedrock/test will not run successfully:
```
[ RUN          ] ChainedHTTPTest::test
*** Error in `../bedrock': corrupted size vs. prev_size: 0x00000000020ea1f0 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7f919e5287e5]
/lib/x86_64-linux-gnu/libc.so.6(+0x80dfb)[0x7f919e531dfb]
/lib/x86_64-linux-gnu/libc.so.6(cfree+0x4c)[0x7f919e53553c]
../bedrock(_ZN11STCPManagerD0Ev+0x29)[0x679ebb]
../bedrock(_ZN13BedrockPluginD1Ev+0xa1)[0x52a4b9]
/vagrant/Bedrock/test/clustertest/testplugin/testplugin.so(_ZN24BedrockPlugin_TestPluginD1Ev+0x56)[0x7f919d516d6c]
/vagrant/Bedrock/test/clustertest/testplugin/testplugin.so(_ZN24BedrockPlugin_TestPluginD0Ev+0x18)[0x7f919d516d88]
../bedrock(_ZN13BedrockServerD2Ev+0xff4)[0x54e8c4]
../bedrock(main+0x4dd6)[0x521604]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f919e4d1830]
../bedrock(_start+0x29)[0x515fa9]
======= Memory map: ========
```
need to dive in.

tylerkaraszewski commented:
`make clean` and rebuild, this is almost certainly a linker issue with an out-of-date library.

righdforsa commented:
> make clean and rebuild, this is almost certainly a linker issue with an out-of-date library.

that's what I thought too, but make clean/make all did not fix the problem. I'll try again

righdforsa commented:
Did make clean and make all again, no love. Checked out master and caught it up, no love. Updated mbedtls submodule with master, no love.

So, it's either broken in master, or my environment is really jacked up. Here's the full error, 100% repeatable on ChainedHTTPTest. Can you try in your environment @tylerkaraszewski ?

```
vagrant@expensidev:/vagrant/Bedrock/test$ ./test -only ChainedHTTP
[--------------]
[ RUN          ] ChainedHTTPTest::test
*** Error in `../bedrock': corrupted size vs. prev_size: 0x000000000153b1f0 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7f0564b617e5]
/lib/x86_64-linux-gnu/libc.so.6(+0x80dfb)[0x7f0564b6adfb]
/lib/x86_64-linux-gnu/libc.so.6(cfree+0x4c)[0x7f0564b6e53c]
../bedrock(_ZN11STCPManagerD0Ev+0x29)[0x6742d9]
../bedrock(_ZN13BedrockPluginD1Ev+0xa1)[0x52a219]
/vagrant/Bedrock/test/clustertest/testplugin/testplugin.so(_ZN24BedrockPlugin_TestPluginD1Ev+0x56)[0x7f0563b4fdbc]
/vagrant/Bedrock/test/clustertest/testplugin/testplugin.so(_ZN24BedrockPlugin_TestPluginD0Ev+0x18)[0x7f0563b4fdd8]
../bedrock(_ZN13BedrockServerD2Ev+0xff4)[0x549a68]
../bedrock(main+0x4dd6)[0x521364]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f0564b0a830]
../bedrock(_start+0x29)[0x515d09]
======= Memory map: ========
00400000-008b5000 r-xp 00000000 00:28 9558980                            /vagrant/Bedrock/bedrock
00ab4000-00ab6000 r--p 004b4000 00:28 9558980                            /vagrant/Bedrock/bedrock
00ab6000-00abb000 rw-p 004b6000 00:28 9558980                            /vagrant/Bedrock/bedrock
00abb000-00abe000 rw-p 00000000 00:00 0 
01527000-01559000 rw-p 00000000 00:00 0                                  [heap]
7f0530000000-7f0530041000 rw-p 00000000 00:00 0 
7f0530041000-7f0534000000 ---p 00000000 00:00 0 
7f0538000000-7f0538041000 rw-p 00000000 00:00 0 
7f0538041000-7f053c000000 ---p 00000000 00:00 0 
7f053c000000-7f053c041000 rw-p 00000000 00:00 0 
7f053c041000-7f0540000000 ---p 00000000 00:00 0 
7f0540000000-7f0540041000 rw-p 00000000 00:00 0 
7f0540041000-7f0544000000 ---p 00000000 00:00 0 
7f0544000000-7f0544041000 rw-p 00000000 00:00 0 
7f0544041000-7f0548000000 ---p 00000000 00:00 0 
7f0548000000-7f0548041000 rw-p 00000000 00:00 0 
7f0548041000-7f054c000000 ---p 00000000 00:00 0 
7f054c000000-7f054c041000 rw-p 00000000 00:00 0 
7f054c041000-7f0550000000 ---p 00000000 00:00 0 
7f0550000000-7f0550041000 rw-p 00000000 00:00 0 
7f0550041000-7f0554000000 ---p 00000000 00:00 0 
7f0554000000-7f05540f7000 rw-p 00000000 00:00 0 
7f05540f7000-7f0558000000 ---p 00000000 00:00 0 
7f055abdc000-7f055abf3000 r-xp 00000000 08:01 2074                       /lib/x86_64-linux-gnu/libresolv-2.23.so
7f055abf3000-7f055adf3000 ---p 00017000 08:01 2074                       /lib/x86_64-linux-gnu/libresolv-2.23.so
7f055adf3000-7f055adf4000 r--p 00017000 08:01 2074                       /lib/x86_64-linux-gnu/libresolv-2.23.so
7f055adf4000-7f055adf5000 rw-p 00018000 08:01 2074                       /lib/x86_64-linux-gnu/libresolv-2.23.so
7f055adf5000-7f055adf7000 rw-p 00000000 00:00 0 
7f055adf7000-7f055adfc000 r-xp 00000000 08:01 2070                       /lib/x86_64-linux-gnu/libnss_dns-2.23.so
7f055adfc000-7f055affc000 ---p 00005000 08:01 2070                       /lib/x86_64-linux-gnu/libnss_dns-2.23.so
7f055affc000-7f055affd000 r--p 00005000 08:01 2070                       /lib/x86_64-linux-gnu/libnss_dns-2.23.so
7f055affd000-7f055affe000 rw-p 00006000 08:01 2070                       /lib/x86_64-linux-gnu/libnss_dns-2.23.so
7f055affe000-7f055afff000 ---p 00000000 00:00 0 
7f055afff000-7f055b7ff000 rw-p 00000000 00:00 0 
7f055b7ff000-7f055b800000 ---p 00000000 00:00 0 
7f055b800000-7f055c000000 rw-p 00000000 00:00 0 
7f055c000000-7f055c021000 rw-p 00000000 00:00 0 
7f055c021000-7f0560000000 ---p 00000000 00:00 0 
7f0560112000-7f0560113000 ---p 00000000 00:00 0 
7f0560113000-7f0560913000 rw-p 00000000 00:00 0 
7f0563118000-7f0563119000 ---p 00000000 00:00 0 
7f0563119000-7f0563919000 rw-p 00000000 00:00 0 
7f0563919000-7f0563924000 r-xp 00000000 08:01 2079                       /lib/x86_64-linux-gnu/libnss_files-2.23.so
7f0563924000-7f0563b23000 ---p 0000b000 08:01 2079                       /lib/x86_64-linux-gnu/libnss_files-2.23.so
7f0563b23000-7f0563b24000 r--p 0000a000 08:01 2079                       /lib/x86_64-linux-gnu/libnss_files-2.23.so
7f0563b24000-7f0563b25000 rw-p 0000b000 08:01 2079                       /lib/x86_64-linux-gnu/libnss_files-2.23.so
7f0563b25000-7f0563b2b000 rw-p 00000000 00:00 0 
7f0563b2b000-7f0563b6c000 r-xp 00000000 00:28 9559290                    /vagrant/Bedrock/test/clustertest/testplugin/testplugin.so
7f0563b6c000-7f0563d6c000 ---p 00041000 00:28 9559290                    /vagrant/Bedrock/test/clustertest/testplugin/testplugin.so
7f0563d6c000-7f0563d6d000 r--p 00041000 00:28 9559290                    /vagrant/Bedrock/test/clustertest/testplugin/testplugin.so
7f0563d6d000-7f0563d6f000 rw-p 00042000 00:28 9559290                    /vagrant/Bedrock/test/clustertest/testplugin/testplugin.so
7f0563d6f000-7f0563d70000 ---p 00000000 00:00 0 
7f0563d70000-7f0564570000 rw-p 00000000 00:00 0 
7f0564570000-7f0564678000 r-xp 00000000 08:01 2085                       /lib/x86_64-linux-gnu/libm-2.23.so
7f0564678000-7f0564877000 ---p 00108000 08:01 2085                       /lib/x86_64-linux-gnu/libm-2.23.so
7f0564877000-7f0564878000 r--p 00107000 08:01 2085                       /lib/x86_64-linux-gnu/libm-2.23.so
7f0564878000-7f0564879000 rw-p 00108000 08:01 2085                       /lib/x86_64-linux-gnu/libm-2.23.so
7f0564879000-7f05648e9000 r-xp 00000000 08:01 59232                      /lib/x86_64-linux-gnu/libpcre.so.3.13.3
7f05648e9000-7f0564ae8000 ---p 00070000 08:01 59232                      /lib/x86_64-linux-gnu/libpcre.so.3.13.3
7f0564ae8000-7f0564ae9000 r--p 0006f000 08:01 59232                      /lib/x86_64-linux-gnu/libpcre.so.3.13.3
7f0564ae9000-7f0564aea000 rw-p 00070000 08:01 59232                      /lib/x86_64-linux-gnu/libpcre.so.3.13.3
7f0564aea000-7f0564caa000 r-xp 00000000 08:01 2082                       /lib/x86_64-linux-gnu/libc-2.23.so
7f0564caa000-7f0564eaa000 ---p 001c0000 08:01 2082                       /lib/x86_64-linux-gnu/libc-2.23.so
7f0564eaa000-7f0564eae000 r--p 001c0000 08:01 2082                       /lib/x86_64-linux-gnu/libc-2.23.so
7f0564eae000-7f0564eb0000 rw-p 001c4000 08:01 2082                       /lib/x86_64-linux-gnu/libc-2.23.so
7f0564eb0000-7f0564eb4000 rw-p 00000000 00:00 0 
7f0564eb4000-7f0564ecb000 r-xp 00000000 08:01 59152                      /lib/x86_64-linux-gnu/libgcc_s.so.1
7f0564ecb000-7f05650ca000 ---p 00017000 08:01 59152                      /lib/x86_64-linux-gnu/libgcc_s.so.1
7f05650ca000-7f05650cb000 r--p 00016000 08:01 59152                      /lib/x86_64-linux-gnu/libgcc_s.so.1
7f05650cb000-7f05650cc000 rw-p 00017000 08:01 59152                      /lib/x86_64-linux-gnu/libgcc_s.so.1
7f05650cc000-7f056529f000 r-xp 00000000 08:01 58384                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
7f056529f000-7f056549e000 ---p 001d3000 08:01 58384                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
7f056549e000-7f05654a9000 r--p 001d2000 08:01 58384                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
7f05654a9000-7f05654ac000 rw-p 001dd000 08:01 58384                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
7f05654ac000-7f05654af000 rw-p 00000000 00:00 0 
7f05654af000-7f05654c8000 r-xp 00000000 08:01 2151                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f05654c8000-7f05656c7000 ---p 00019000 08:01 2151                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f05656c7000-7f05656c8000 r--p 00018000 08:01 2151                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f05656c8000-7f05656c9000 rw-p 00019000 08:01 2151                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f05656c9000-7f05656e1000 r-xp 00000000 08:01 2068                       /lib/x86_64-linux-gnu/libpthread-2.23.so
7f05656e1000-7f05658e0000 ---p 00018000 08:01 2068                       /lib/x86_64-linux-gnu/libpthread-2.23.so
7f05658e0000-7f05658e1000 r--p 00017000 08:01 2068                       /lib/x86_64-linux-gnu/libpthread-2.23.so
7f05658e1000-7f05658e2000 rw-p 00018000 08:01 2068                       /lib/x86_64-linux-gnu/libpthread-2.23.so
7f05658e2000-7f05658e6000 rw-p 00000000 00:00 0 
7f05658e6000-7f05658ef000 r-xp 00000000 08:01 60377                      /usr/lib/x86_64-linux-gnu/libpcrecpp.so.0.0.1
7f05658ef000-7f0565aee000 ---p 00009000 08:01 60377                      /usr/lib/x86_64-linux-gnu/libpcrecpp.so.0.0.1
7f0565aee000-7f0565aef000 r--p 00008000 08:01 60377                      /usr/lib/x86_64-linux-gnu/libpcrecpp.so.0.0.1
7f0565aef000-7f0565af0000 rw-p 00009000 08:01 60377                      /usr/lib/x86_64-linux-gnu/libpcrecpp.so.0.0.1
7f0565af0000-7f0565af3000 r-xp 00000000 08:01 2064                       /lib/x86_64-linux-gnu/libdl-2.23.so
7f0565af3000-7f0565cf2000 ---p 00003000 08:01 2064                       /lib/x86_64-linux-gnu/libdl-2.23.so
7f0565cf2000-7f0565cf3000 r--p 00002000 08:01 2064                       /lib/x86_64-linux-gnu/libdl-2.23.so
7f0565cf3000-7f0565cf4000 rw-p 00003000 08:01 2064                       /lib/x86_64-linux-gnu/libdl-2.23.so
7f0565cf4000-7f0565d1a000 r-xp 00000000 08:01 2067                       /lib/x86_64-linux-gnu/ld-2.23.so
7f0565f02000-7f0565f0a000 rw-p 00000000 00:00 0 
7f0565f18000-7f0565f19000 rw-p 00000000 00:00 0 
7f0565f19000-7f0565f1a000 r--p 00025000 08:01 2067                       /lib/x86_64-linux-gnu/ld-2.23.so
7f0565f1a000-7f0565f1b000 rw-p 00026000 08:01 2067                       /lib/x86_64-linux-gnu/ld-2.23.so
7f0565f1b000-7f0565f1c000 rw-p 00000000 00:00 0 
7fff2b2f6000-7fff2b317000 rw-p 00000000 00:00 0                          [stack]
7fff2b3a5000-7fff2b3a7000 r--p 00000000 00:00 0                          [vvar]
7fff2b3a7000-7fff2b3a9000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
[       PASSED ] ChainedHTTPTest::test
[--------------]

[==============]
[ TEST RESULTS ] Passed: 1, Failed: 0
[==============]
vagrant@expensidev:/vagrant/Bedrock/test$ 
```

righdforsa commented:
FWIW, technically, the test "passes", so it's possible it's been broken for a while, but I didn't see it in my previous tests.

righdforsa commented:
Going through a git bisect on this.

coleaeason commented:
My guess is this is something in your vm

righdforsa commented:
```
b0b7cce83429ddaf0534bfb4ace5297060f8580c is the first bad commit
commit b0b7cce83429ddaf0534bfb4ace5297060f8580c
Author: Tyler Karaszewski <tyler@expensify.com>
Date:   Tue Mar 31 10:30:05 2020 -0700

    Fix loglevels

:040000 040000 166bd71509ea2d998a4b204a85a9fbc3fd316d07 409993a819f9ec9768373ca475faeeba175b40c6 M	libstuff
:040000 040000 caffbd4cbdfe52312a4e692c985af2b1921c2609 ebec44ca48b6a2568b8a2203fae03808e2e6994b M	sqlitecluster
```

tylerkaraszewski commented:
have you tried restarting the VM?

tylerkaraszewski commented:
I just reproduced this as well. Working on it.

righdforsa commented:
Ok, now that we've confirmed, this PR is ready to go.

righdforsa commented:
Confirmed Bedrock/test tests complete with Tyler's fix in.

On Tue, 14 Apr 2020 at 18:14, Christopher Peters <notifications@github.com>
wrote:

> Ok, now that we've confirmed, this PR is ready to go.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Expensify/Bedrock/pull/733#issuecomment-613598773>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMUKP6XKENZ2XEDSNX5T3LRMSRYNANCNFSM4LHS6FKQ>
> .
>

righdforsa commented:
@tylerkaraszewski @coleaeason  rerunning travis on the PR check. Code should now be ready to go. Any chance you guys can review today? No rush, just eager.

coleaeason commented on sqlitecluster/SQLiteNode.cpp:
```diff
@@ -561,14 +564,13 @@ bool SQLiteNode::update() {
+            _priority > 0 && _priority > highestPriorityPeer->calc("Priority")) {
             // Yep -- time for us to stand up -- clear everyone's
             // last approval status as they're about to send them.
-            SASSERT(_priority > 0); // Permafollower should never stand up
```
you should be able to keep this assert right?

righdforsa replied:
There's no reason to keep it with the update to the conditional in line 571.

coleaeason commented on test/clustertest/tests/MultipleLeaderSyncTest.cpp:
```diff
@@ -0,0 +1,95 @@
+                              TEST(MultipleLeaderSyncTest::test)
+                             ) { }
+
+    void setup() {}
```
how come you aren't using setup or tear down? all other tests use these methods to create and clean up the tester and other variables, would be nice if your tests followed suit

righdforsa replied:
Tyler requested that I not use setup and teardown if they're simple, because without them, the tester gets destroyed when it goes out of scope. i'm not sure that one is obviously better. @tylerkaraszewski ?

tylerkaraszewski replied:
My view is that if you don't need separate `setup` and `teardown` functions, you can do this:

```
BedrockTester t(some params);
t.whatever();
// done.
```

But if you insist on using them then you need:
```
BedrockTester* t; // Needs to be a pointer because there's no default object for a BedrockTester

setup() {
    t = new Bedrocktester(params); // now we've dynamically allocated it so we need to explicitly clean it up.
}

test () {
    t->whatever(); // If this throws does our object get deleted (I think so)?
}

teardown() {
    delete t; // Wouldn't need this if we just made it on the stack.
}
```

coleaeason replied:
can we remove the setup/teardown functions then? or does the test library require them to exist and be no ops?

(I actually like the scoped objects better, it just doesn't match most of our tests)

tylerkaraszewski replied:
Yes, we can remove them from this test, they don't need to exist, they're optional.

coleaeason commented on test/clustertest/tests/PermafollowerTest.cpp:
```diff
@@ -0,0 +1,56 @@
+                              TEST(PermafollowerTest::test)
+                             ) { }
+
+    void setup() {}
```
same question

righdforsa replied:
same answer 😃

coleaeason commented on test/lib/BedrockTester.cpp:
```diff
@@ -552,35 +559,66 @@ bool BedrockTester::waitForStates(set<string> states, uint64_t timeoutUS)
+    // we timed out, let's get some debugging info
+    string stateString;
+    auto it = states.begin();
+    while (it != states.end()) {
```
use `SComposeList(states)` to get a string from the set

righdforsa replied:
nice, thanks

coleaeason commented on BedrockServer.cpp:
```diff
@@ -364,11 +364,49 @@ void BedrockServer::sync(const SData& args,
+            // If we were LEADING, but we've transitioned, then something's gone wrong (perhaps we got disconnected
+            // from the cluster).
+            if ((preUpdateState == SQLiteNode::LEADING || preUpdateState == SQLiteNode::STANDINGDOWN)
+                     && nodeState != SQLiteNode::LEADING && nodeState != SQLiteNode::STANDINGDOWN) {
```
👍 to tyler's comment

righdforsa replied:
I agree that the conditional could be different, but I don't follow the suggestion of `nodeState == SQLiteNode::STANDINGDOWN`. nodeState can never by STANDINGDOWN there, because of the conditional at 372. I would propose removing the `nodeState !=` conditionals on line 377, and keeping the `preUpdateState` conditionals on line 376.

tylerkaraszewski replied:
Oops, I meant `nodeState == SQLiteNode::FOLLOWING`, copy/pasted wrong line.

coleaeason commented on test/lib/BedrockTester.cpp:
```diff
@@ -552,35 +559,66 @@ bool BedrockTester::waitForStates(set<string> states, uint64_t timeoutUS)
+        it++;
+    }
+    string statusString;
+    for (auto& item : json) {
```
use `SComposeJSONObject(json)` to just get json string that you can print

coleaeason commented on test/lib/BedrockTester.cpp:
```diff
@@ -552,35 +559,66 @@ bool BedrockTester::waitForStates(set<string> states, uint64_t timeoutUS)
 
-    return 1;
+            // if the value matches, return, otherwise wait
+            if (result != "" && result == testValue) {
```
why the empty string check here? what if you want the test value to be an empty string?

righdforsa commented on BedrockServer.cpp:
```diff
@@ -430,31 +450,43 @@ void BedrockServer::sync(const SData& args,
+                    continue;
                 }
-                continue;
+            } catch(const std::system_error& e){
```
I think we want to keep this. If we choose not to keep it, we'll still crash and get a stack trace. With it, we get a little extra clarity. I'm open to removing if you prefer.

coleaeason replied:
I don't think we want to keep it. We don't really get extra clarity, the stack is the same either way.

coleaeason commented on test/lib/PortMap.h:
```diff
@@ -0,0 +1,84 @@
+
+    // Waits for a particular port to be free to bind to. This is useful when we've killed a server, because sometimes
+    // it takes the OS a few seconds to make the port available again.
+    int waitForPort(uint16_t port) {
```
why is this an  int instead of a bool? only ever returns 0 or 1

righdforsa replied:
I kept it from the previous implementation. I assume it's an int because the bind function returns 0 for success and 1 for failure, which is the opposite of the truth value of those. It could be done, just seemed non-trivial to replace the logic in BedrockTester that relies on this. I agree, it's confusing though.

coleaeason commented on test/lib/PortMap.h:
```diff
@@ -0,0 +1,84 @@
+        return 1;
+    }
+
+    PortMap(uint16_t from = START_PORT) : _from(from)
```
move the constructor to the top of the class definition, above al of the functions.

coleaeason commented on test/lib/PortMap.h:
```diff
@@ -0,0 +1,84 @@
+    static const int64_t MAX_PORT = 20000;
+
+    // Get an unused port
+    uint16_t getPort()
```
since you moved this class to it's own file it makes sense to move these into their own implementation file, there's no reason they need to be in the header.

righdforsa commented on BedrockServer.cpp:
```diff
@@ -380,10 +418,18 @@ void BedrockServer::sync(const SData& args,
+                try {
+                    SDEBUG("Locking the _syncThreadCommitMutex for _upgradeDB");
+                    server._syncThreadCommitMutex.lock();
+                } catch(const std::system_error& e){
```
This allows us to capture the system error code, and makes the failure clearer IMO, but I'm happy to remove it since the behavior is the same and we'll still get a stack trace.

righdforsa commented on BedrockServer.cpp:
```diff
@@ -430,31 +450,43 @@ void BedrockServer::sync(const SData& args,
+                    continue;
                 }
-                continue;
+            } catch(const std::system_error& e){
```
when it threw in testing, the stack trace pointed to the closure of the `if` block, if I recall correctly, which is why I wrapped the whole block.

righdforsa commented on BedrockServer.cpp:
```diff
@@ -567,11 +613,13 @@ void BedrockServer::sync(const SData& args,
 
                         // Move on to the next command until this one finishes.
                         core.rollback();
+                        SDEBUG("Unlocking the _syncThreadCommitMutex after new https request");
```
Can you explain why we should take them out? They were useful in pinpointing the problem with the mutex, which I'm not confident is permanently resolved.

tylerkaraszewski replied:
If we left in every debug logline that was used to debug a weird race condition, there's literally be 5,000 debug lines in the form of `TYLER this thing happened` littering the entire code base.

righdforsa replied:
So, how do you decide whether a debug flag is useful to keep? The crashes that led to this investigation were easy enough to debug, but finding where the mutex had previously been locked was not. On the other hand, I think we're likely to have fixed these lock issues by moving the code block that handles these failures, so it might make sense to wait. I'm ambivalent, but lean towards verbose logging. Does it matter how many debug log lines there are?

tylerkaraszewski replied:
Currently there are only 39 of them and only a handful touched in 2020, and I think most of those are changes to existing lines, not new ones added. Which is to say we barely use them currently.

Loglines that are only recorded in dev are only useful in dev, in which case they're quite easy to add when debugging a specific problem.

Like I said, we'd have thousands of these if I kept every logline I added while debugging every issue. At what point does the signal/noise ratio get so low that we rip out all of the debug log lines for the last 30 bugs that we investigated? How would you analyze the log lines you added here if every mutex lock/unlock logged a debug line?

```
65309862 (Tyler Karaszewski 2017-09-08 08:09:00 -0700 246)         SDEBUG("Received command #" << (int)packet.sequenceID << ": '" << SToHex(packet.serialize()) << "'");
9ce19c33 (Tyler Karaszewski 2020-02-26 14:39:30 -0800 1675)             SDEBUG("Plugin " << pair.first << " handling command " << command->request.methodLine);
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700  96)                 SDEBUG("Connect to '" << socket->addr << "' failed with SO_ERROR #" << result << ", closing.");
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700 103)             SDEBUG("Connect to '" << socket->addr << "' succeeded.");
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700 119)                 SDEBUG("sslState=" << SSSLGetState(socket->ssl) << ", canrecv=" << SFDAnySet(fdm, socket->s, SREADEVTS)
64c3c924 (Tyler Karaszewski 2017-03-16 12:11:37 -0700 141)                 SDEBUG("Connection to '" << socket->addr << "' died (recv=" << aliveAfterRecv << ", send="
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700 160)                         SDEBUG("Graceful shutdown of SSL socket '" << socket->addr << "'");
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700 186)                         SDEBUG("Graceful shutdown of socket '" << socket->addr << "'");
64c3c924 (Tyler Karaszewski 2017-03-16 12:11:37 -0700 205)     SDEBUG("Shutting down socket '" << socket->addr << "' (" << how << ")");
0214b0b2 (Tyler Karaszewski 2019-01-18 09:51:40 -0400 213)     SDEBUG("Closing socket '" << socket->addr << "'");
ecda922d (Tyler Karaszewski 2016-10-14 11:17:40 -0700 62)             SDEBUG("Accepting socket from '" << addr << "' on port '" << port.host << "'");
c2e92007 (Tyler Karaszewski 2017-09-07 18:16:16 -0700 170)                 SDEBUG("Incoming connection failed from '" << socket->addr << "' (" << e.what() << "), empty buffers");
c2e92007 (Tyler Karaszewski 2017-09-07 18:16:16 -0700 72)         SDEBUG("Failed to deserialize JSON-encoded SQResult (" << e.what() << "): " << json);
55608659 (Ionatan Wiznia    2018-03-14 14:08:33 +0100 278) #define SDEBUG(_MSG_) SSYSLOG(LOG_DEBUG, "[dbug] " << SLOGPREFIX << _MSG_)
b0b7cce8 (Tyler Karaszewski 2020-03-31 10:30:05 -0700   4) #define PDEBUG(_MSG_) SDEBUG("->{" << peer->name << "} " << _MSG_)
551cbbea (Tyler Karaszewski   2016-09-29 13:53:22 -0700 1253)                 SDEBUG("Expecting: '" << _CH_ << "', found 'eol'");                                                    \
551cbbea (Tyler Karaszewski   2016-09-29 13:53:22 -0700 1255)                 SDEBUG("Expecting: '" << _CH_ << "', found '" << ptr << "'");                                          \
551cbbea (Tyler Karaszewski   2016-09-29 13:53:22 -0700 1352)             SDEBUG("Parsed: '" << name << "':'" << value << "'");
b4989cf7 (Tyler Karaszewski   2017-02-16 11:59:21 -0800 2449)         SDEBUG(sql);
551cbbea (Tyler Karaszewski  2016-09-29 13:53:22 -0700  581)         SDEBUG("Connected to " << numLoggedInFullPeers << " of " << numFullPeers << " full peers (" << peerList.size()
1aa76dd7 (Tyler Karaszewski  2017-05-08 16:34:46 -0700  715)                             SDEBUG("Peer '" << peer->name << "' has approved transaction.");
3c76b6aa (Tyler Karaszewski  2017-03-20 12:19:04 -0700  751)             SDEBUG(     "numFullPeers="           << numFullPeers
c918f743 (Tyler Karaszewski  2018-01-16 16:14:09 -0800  782)                 SDEBUG("Committing current transaction because consistentEnough: " << _db.getUncommittedQuery());
a29fc8c2 (Mike Brannigan     2018-04-27 11:35:21 -0700 1795)         SDEBUG("Setting state timeout of " << timeout / 1000 << "ms");
c918f743 (Tyler Karaszewski  2018-01-16 16:14:09 -0800 1989)         SDEBUG("Committing current transaction because _recvSynchronize: " << _db.getUncommittedQuery());
5646707b (Tyler Karaszewski  2020-03-11 14:29:12 -0700 2277)     SDEBUG("Committing current transaction because COMMIT_TRANSACTION: " << _db.getUncommittedQuery());
8f568f10 (Tyler Karaszewski 2020-03-25 12:56:10 -0700  276)         SDEBUG("[checkpoint] starting thread with count: " << object->_sharedData->_currentPageCount.load());
006d1244 (Tyler Karaszewski 2017-01-24 18:24:42 -0300  416)     SDEBUG("Beginning transaction");
f901a3ec (Tyler Karaszewski 2017-01-25 13:04:18 -0300  447)     SDEBUG("[concurrent] Beginning transaction");
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700  711)     SDEBUG("Prepared transaction");
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700  745)     SDEBUG("Committing transaction");
4e3415f8 (Tyler Karaszewski 2018-01-31 12:40:06 -0800  777)         SDEBUG("Commit successful (" << _sharedData->_commitCount.load() << "), releasing commitLock.");
551cbbea (Tyler Karaszewski 2016-09-29 13:53:22 -0700  919)     SDEBUG("Getting commits #" << fromIndex << "-" << toIndex);
56d3b882 BedrockCore.cpp (Tyler Karaszewski 2020-01-30 15:32:14 -0800  78)         SDEBUG("Peeking at '" << request.methodLine << "' with priority: " << command->priority);
49fd663a BedrockCore.cpp (Tyler Karaszewski 2020-02-26 14:42:31 -0800  95)             SDEBUG("Plugin '" << command->getName() << "' peeked command '" << request.methodLine << "'");
73999b2b BedrockCore.cpp (Cole              2018-10-18 06:49:36 -0700 181)         SDEBUG("Processing '" << request.methodLine << "'");
49fd663a BedrockCore.cpp (Tyler Karaszewski 2020-02-26 14:42:31 -0800 205)                 SDEBUG("Plugin '" << command->getName() << "' processed command '" << request.methodLine << "'");
551cbbea (Tyler Karaszewski           2016-09-29 13:53:22 -0700 285)         SDEBUG("Resetting database");
551cbbea (Tyler Karaszewski           2016-09-29 13:53:22 -0700 293)         SDEBUG("Verifying database exists");
```

righdforsa commented on sqlitecluster/SQLiteNode.cpp:
```diff
@@ -561,14 +564,13 @@ bool SQLiteNode::update() {
+        // enough full peers to achieve quorum, we should be leader.
         if (!currentLeader && numLoggedInFullPeers * 2 >= numFullPeers &&
-            _priority > highestPriorityPeer->calc("Priority")) {
+            _priority > 0 && _priority > highestPriorityPeer->calc("Priority")) {
```
If you're not checking for `_priority > 0`, a priority 0 permaslave in waiting will stand up if the priorities of the full peers are -1 because they haven't switched to using their real priority.

tylerkaraszewski replied:
If `_priority > 0` is true then `_priority` is true, right?

tylerkaraszewski replied:
Yes, I can't read.

Can `highestPriorityPeer->calc("Priority")` be less than 0?

righdforsa replied:
It can be less than zero, because of the "-1" fake priority for hosts that haven't reached WAITING yet.

> Would that make numLoggedInFullPeers * 2 >= numFullPeers fail?
Not sure, I will check. I think there's something about "assuming they're logged in if they aren't connected" or something. I'll dig in and get back to you.

righdforsa replied:
The way we set `numLoggedInFullPeers` is by looking to see if the peer is logged in. `SIEquals((*peer)["LoggedIn"], "true")` This has a race condition, right, if you lose connection with half your peers at the same time?

righdforsa replied:
`numLoggedInFullPeers` first excludes permafollowers (based on configured peerList) and then depends on the logged in state:
```
            // Make sure we're a full peer
            if (peer->params["Permafollower"] != "true") {
                // Verify we're logged in
                ++numFullPeers;
                if (SIEquals((*peer)["LoggedIn"], "true")) {
                    // Verify we're still fresh
                    ++numLoggedInFullPeers;
```

so, any node can be logged in, but still advertising the "fake" priority.

righdforsa commented on BedrockServer.cpp:
```diff
@@ -364,11 +364,49 @@ void BedrockServer::sync(const SData& args,
-        // If the node's not in a ready state at this point, we'll probably need to read from the network, so start the
-        // main loop over. This can let us wait for logins from peers (for example).
+        // If the node's not in a ready state at this point, we'll probably need to read from the network. Test
+        // conditions from transitional states and then start the main loop over. This can let us wait for logins
```
I'm open to improvements, but "so start the main loop over" doesn't make sense if we're about to drop into a bunch of additional code.

tylerkaraszewski replied:
Should `network. Test` be `network, test`, what do you mean by "transitional states"?

righdforsa replied:
Comma or not just looks like a writing style choice, IMO. I'll make that change if it resolves your concern. Alternatively, if i moved "Test" to its own line, my meaning might be a little clearer, as in "this is what is the code does, which is separate from the "why" we do this in the previous sentence.

By transitional states, I was attempting to generalize what "had a commit interrupted because of losing quorum" represents. Not very clear, I agree. Pushing a variation, LMK what you think.

righdforsa commented on BedrockServer.cpp:
```diff
@@ -567,11 +613,13 @@ void BedrockServer::sync(const SData& args,
                     }
 
                     // If we just started a new HTTPS request, save it for later.
+                    // WTF can this ever be reached? maybe peekCommand creates an httpsRequest?
```
I was confused and answered it for myself. will remove

righdforsa commented:
Incorporated feedback and re-tested. Passed `clustertest -repeatCount=10` 2 out of 4 times. (the other times had 1 failure each) Two outstanding comments, waiting on you guys.

righdforsa commented:
@tylerkaraszewski @coleaeason Two remaining outstanding comments threads, ready for re-review.

righdforsa commented:
🎉 woo-hoo! Tyler approved. That leaves you @coleaeason . Make my dreams come true!

coleaeason commented:
annnnnnnd there is a single unsigned commit in the middle of this pr

righdforsa commented:
Closing in favor of https://github.com/Expensify/Bedrock/pull/733
